kong/router/compat.lua

```diff
@@ -295,8 +302,58 @@ local function get_exp_and_priority(route)
 end
 
 
-function _M.new(routes, cache, cache_neg, old_router)
-  return atc.new(routes, cache, cache_neg, old_router, get_exp_and_priority)
+-- group array-like table t by the function f, returning a table mapping from
+-- the result of invoking f on one of the elements to the actual elements.
+local function group_by(t, f)
+  local result = {}
+  for _, value in ipairs(t) do
+    local key = f(value)
+    if result[key] then
+      result[key][#result[key] + 1] = value
+    else
+      result[key] = {value}
+    end
+  end
+  return result
+end
+
+-- split routes into multiple routes, one for each prefix length and one for all
+-- regular expressions
+local function split_route_by_path_into(rs, split_rs)
+  if is_empty_field(rs.route.paths) or #rs.route.paths == 1 then
+    split_rs[#split_rs + 1] = rs
+    return
+  end
+
+  assert(rs.route.paths)
```

Comment:
This looks to be guarded against by `is_empty_field(rs.route.paths)` in the clause on line 323. If that weren't the case then the second half of the conditional (`or #rs.route.paths == 1`) would be "dangerous" due to not performing a nil check before attempting to read the table length.